pkg/scheduler/queue/tenant_queues.go

```diff
@@ -203,23 +171,12 @@ func (qb *queueBroker) dequeueRequestForQuerier(
 		tenant = qb.tenantQuerierAssignments.tenantsByID[tenantID]
 	}
 
-	// TODO (casie): When deprecating TreeQueue, clean this up.
-	// This cannot be handled by the Tree interface without defining some other, more expansive interfaces
-	// between the legacy and integrated tree queues, which would be more overhead than it's worth, given
-	// that we will eventually retire the legacy tree queue.
-	if tq, ok := qb.tree.(*TreeQueue); ok {
-		queueNodeAfterDequeue := tq.getNode(queuePath)
-		if queueNodeAfterDequeue == nil {
-			// queue node was deleted due to being empty after dequeue
-			qb.tenantQuerierAssignments.removeTenant(tenant.tenantID)
-		}
-	} else if itq, ok := qb.tree.(*MultiQueuingAlgorithmTreeQueue); ok {
-		queueNodeAfterDequeue := itq.GetNode(queuePath)
-		if queueNodeAfterDequeue == nil && len(qb.tenantQuerierAssignments.tenantNodes[string(tenantID)]) == 0 {
-			// queue node was deleted due to being empty after dequeue
-			qb.tenantQuerierAssignments.removeTenant(tenantID)
-		}
+	queueNodeAfterDequeue := qb.tree.GetNode(queuePath)
+	if queueNodeAfterDequeue == nil && len(qb.tenantQuerierAssignments.tenantNodes[string(tenantID)]) == 0 {
+		// queue node was deleted due to being empty after dequeue
+		qb.tenantQuerierAssignments.removeTenant(tenantID)
```

Comment:
Related to making `GetNode` part of the public tree interface:
not really worth tackling in this PR, but as we have moved more of the node state management back into the tree, I find myself more and more wanting to have the tree just delete the empty and report to the queueing algorithm that it did so - then each algorithm can update its internal state but not actually do the deleting.
Deleting the empty node is generally assumed to happen by all our implementations and interface dosctrings, but still has to be done by each algorithm implementation separately.

This is related to `GetNode` because this above is the only non-test usage of it, which could be eliminated if we just had the tree report that the node was deleted instead.
`GetNode` is a great thing to have for tests, but I don't really know if it belongs in the public interface in the long term, since we want all interactions from consumers of the interface to start at the root of the tree to work as expected.

Reply:
I think this'll need a bit more discussion, yeah. I agree with some of your points, but no part of the tenantQuerierAssignments' `QueuingAlgorithm` implementation currently touches `tenantsByID`, and having it do so would (I think) greatly complicate the task of separating the many responsibilities `tenantQuerierAssignments` currently holds.

Reply:
:+1: I sort of conflated two concepts:
1. Not making deleting the node the queue algorithm's responsibility
2. Reporting via `tree.Dequeue` return values to the caller that the queue path is now empty/gone so that the caller can do things `qb.tenantQuerierAssignments.removeTenant(tenantID)` without using the returned `queuePath` to reach down into the tree via `GetNode`, check for existence, then doing `removeTenant`